### PR TITLE
Moved ProductProvenanceRetriever to FWCore/Framework

### PR DESCRIPTION
**Pull request by Dr15Jones**
#### PR description:

Broke ProductProvenanceRetriever into two parts and left base interface ProductProvenanceLookup in DataFormats/Provenance. This avoids having DataFormats/Provenance depend upon FWCore/Concurrency.

#### PR validation:

The code compiles and the framework unit tests pass.

solves makortel/framework#141

### CODE REVIEW

**Comment by Dr15Jones** [comment]
please test

**Comment by Dr15Jones** [comment]
@wddgit please review

**Comment by Dr15Jones** [comment]
@davidlange6 I hope this will help the MODULES build

File: DataFormats/Provenance/src/ProductProvenanceLookup.cc [hunk 1]
```diff
@@ -0,0 +1,117 @@
+#include "DataFormats/Provenance/interface/ProductProvenanceLookup.h"
+#include "DataFormats/Provenance/interface/ProductRegistry.h"
+#include "FWCore/Utilities/interface/EDMException.h"
+
+#include <cassert>
+#include <iostream>
+#include <limits>
+
```
Comment (wddgit) [inline DataFormats/Provenance/src/ProductProvenanceLookup.cc:8]: includes, looks like algorithm is needed and iostream and limits are not

File: FWCore/Framework/src/ProductProvenanceRetriever.cc [hunk 2]
```diff
@@ -0,0 +1,91 @@
+#include "FWCore/Framework/interface/ProductProvenanceRetriever.h"
+#include "DataFormats/Provenance/interface/ProductRegistry.h"
+#include "FWCore/Utilities/interface/EDMException.h"
+
+#include <cassert>
+#include <iostream>
+#include <limits>
+
```
Comment (wddgit) [inline FWCore/Framework/src/ProductProvenanceRetriever.cc:8]: looks like iostream not needed

**Review by wddgit — COMMENTED** [review]
+1
Looks good to me. Seems like it should give identical behavior when running with these changes. I just noticed a few minor include and forward declaration issues.

File: FWCore/Framework/interface/ProductProvenanceRetriever.h [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+/*
+  ProductProvenanceRetriever
+*/
+
+namespace edm {
+  class ProvenanceReaderBase;
+  class ModuleCallingContext;
+  class ProductRegistry;
+
```
[the remark this answers is not in the document]
Comment (Dr15Jones) [reply FWCore/Framework/interface/ProductProvenanceRetriever.h:25]: done

File: DataFormats/Provenance/src/ProductProvenanceLookup.cc (hunk 1 above)
Comment (Dr15Jones) [reply DataFormats/Provenance/src/ProductProvenanceLookup.cc:6]: done

File: FWCore/Framework/src/ProductProvenanceRetriever.cc (hunk 2 above)
Comment (Dr15Jones) [reply FWCore/Framework/src/ProductProvenanceRetriever.cc:8]: done

**Comment by Dr15Jones** [comment]
please test

**Comment by Dr15Jones** [comment]
please test
Given the use of different files from the base line, let's try it again.

**Comment by Dr15Jones** [comment]
please test

the previous failures were also seen in the IBs.

**Comment by Dr15Jones** [comment]
please test

**Comment by Dr15Jones** [comment]
+1

**Comment by qliphy** [comment]
+1

**Pull request merged**